aten/src/ATen/native/cuda/DepthwiseConv3d.cu

```diff
@@ -0,0 +1,634 @@
+      TORCH_CHECK(grad_output.size(i) == expected_output_size[i],
+                  "Expect grad_output to be of same shape as output, got ",
+                  grad_output.size(i), " and ", expected_output_size[i],
+                  " at dimension ", i);
+    }
+  }
+}
+
+}
+
+#define TENSOR_ARG(name, idx) TensorArg name##_arg{ name, #name, idx };
+
+#define NODEF_OR_EQUAL(x, y) ((y) < 0 || (x) == (y))
+#define NODEF_OR_EQUAL_3(x, y1, y2, y3) \
+  (NODEF_OR_EQUAL(x[0], y1) && \
+   NODEF_OR_EQUAL(x[1], y2) && \
+   NODEF_OR_EQUAL(x[1], y3))
+
+#define DWCONV3D_FORWARD_DISPATCH_SPECIALIZATION(kt, kh, kw, dilt, dilh, dilw) \
+  if (NODEF_OR_EQUAL_3(kernel_size, (kt), (kh), (kw)) &&                    \
+      NODEF_OR_EQUAL_3(dilation, (dilt), (dilh), (dilw))) {                 \
+    using accscalar_t = acc_type<scalar_t, true>;                           \
+    conv_depthwise3d_cuda_kernel                                            \
+    <scalar_t, accscalar_t, (kt), (kh), (kw), (dilt), (dilh), (dilw)>       \
+      <<<grid, block, (smem), at::cuda::getCurrentCUDAStream()>>>(          \
+        input_.packed_accessor32<scalar_t, 5>(),                            \
+        output_.packed_accessor32<scalar_t, 5>(),                           \
+        weight_.packed_accessor32<scalar_t, 5>(),                           \
+        bias_ptr,                                                           \
+        stride[0], stride[1], stride[2],                                    \
+        padding[0], padding[1], padding[2],                                 \
+        dilation[0], dilation[1], dilation[2]);                             \
+  } else
+
+#define DWCONV3D_FORWARD_DISPATCH_OTHERS \
+  {                                      \
+    using accscalar_t = acc_type<scalar_t, true>;                           \
+    conv_depthwise3d_cuda_kernel                                            \
+    <scalar_t,accscalar_t, -1, -1, -1, -1, -1, -1>                          \
+      <<<grid, block, (smem), at::cuda::getCurrentCUDAStream()>>>(          \
+        input_.packed_accessor32<scalar_t, 5>(),                            \
+        output_.packed_accessor32<scalar_t, 5>(),                           \
+        weight_.packed_accessor32<scalar_t, 5>(),                           \
+        bias_ptr,                                                           \
+        stride[0], stride[1], stride[2],                                    \
+        padding[0], padding[1], padding[2],                                 \
+        dilation[0], dilation[1], dilation[2]);                             \
+  }
+
+Tensor conv_depthwise3d_cuda(
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    const Tensor& bias,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation) {
+  TENSOR_ARG(input, 1);
+  TENSOR_ARG(weight, 2);
+
+  checkAllSameGPU("conv_depthwise3d_cuda", {input_arg, weight_arg});
```

Comment:
it looks like you need TensorArgs only to call checkAllSameGPU, but wrapping Tensors in TensorArgs is relatively expensive. You can directly compare devices on Tensors.